mpcs/src/basefold/query_phase.rs

```diff
@@ -157,239 +134,172 @@ pub fn batch_verifier_query_phase<E: ExtensionField, Spec: BasefoldSpec<E>>(
-                .by_ref()
-                .peeking_take_while(|(num_vars, _)| **num_vars == cur_num_var)
-                .map(|(_, index)| {
-                    let (lo, hi) = &base_codeword_lo_hi[*index];
-                    let coeff =
-                        <Spec::EncodingScheme as EncodingScheme<E>>::verifier_folding_coeffs(
-                            vp,
-                            cur_num_var
-                                + <Spec::EncodingScheme as EncodingScheme<E>>::get_rate_log()
-                                - 1,
-                            idx,
-                        );
-                    codeword_fold_with_challenge(&[*lo, *hi], *r, coeff, inv_2)
-                })
-                .sum::<E>();
+                // first folding challenge
+                let r = fold_challenges.first().unwrap();
+                let coeff = S::verifier_folding_coeffs(vp, log2_height, idx);
+                let (lo, hi) = reduced_openings[&log2_height];
+                let mut folded = codeword_fold_with_challenge(&[lo, hi], *r, coeff, inv_2);
 
-            let mut n_d_i = n_d_next;
-            for (
-                (pi_comm, r),
-                CommitPhaseProofStep {
-                    sibling_value: leaf,
-                    opening_proof: proof,
-                },
-            ) in commits
-                .iter()
-                .zip_eq(fold_challenges.iter().skip(1))
-                .zip_eq(opening_ext)
-            {
-                cur_num_var -= 1;
+                for (
+                    (pi_comm, r),
+                    CommitPhaseProofStep {
+                        sibling_value,
+                        opening_proof: proof,
+                    },
+                ) in proof
+                    .commits
+                    .iter()
+                    .zip_eq(fold_challenges.iter().skip(1))
+                    .zip_eq(opening_ext)
+                {
+                    cur_num_var -= 1;
+                    log2_height -= 1;
 
-                let is_interpolate_to_right_index = (idx & 1) == 1;
-                let new_involved_codewords = folding_sorted_order_iter
-                    .by_ref()
-                    .peeking_take_while(|(num_vars, _)| **num_vars == cur_num_var)
-                    .map(|(_, index)| {
-                        let (lo, hi) = &base_codeword_lo_hi[*index];
-                        if is_interpolate_to_right_index {
-                            *hi
-                        } else {
-                            *lo
-                        }
-                    })
-                    .sum::<E>();
+                    let idx_sibling = idx & 0x01;
+                    let mut leafs = vec![*sibling_value; 2];
+                    leafs[idx_sibling] = folded;
+                    if let Some((lo, hi)) = reduced_openings.get(&log2_height) {
+                        leafs[idx_sibling] += if idx_sibling == 1 { *hi } else { *lo };
```

Comment:
@hero78119 This is the place that we absorb smaller codeword in the folding loop.